### PR TITLE
Echo Server with TDD implementation

### PR DESCRIPTION
**Pull request by ikarabulut**
This PR covers the first two acceptance criteria for the Echo Server Story
**When a client sends a message to the server, the server sends a response back to the client containing the original message.**
- This is accomplished and tested through the `ServerIO` class. 
In this class, a Socket writer is created using PrintWriter and the sockets output Stream. A Socket Reader is created using BufferedReader which uses the sockets inputStream. I used a BufferedReader since it will read the entire object at once rather than a Scanner that reads char by char.

**A client can send multiple messages to the server and get the echoed response back each time.**
- This is accomplished through the `EchoServer` class.
In this class, an endpoint is established to create the writer and reader that will be used in that server-client connection. From there you can `beginEcho` which runs a while loop taking and printing input until "." is passed when it will break the loop and end the program.

**Multiple clients can send messages to the server and get back their proper responses. - has not been reached in this version**

### CODE REVIEW

File: .idea/inspectionProfiles/Project_Default.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<component name="InspectionProjectProfileManager">
```
Comment (jramnani) [inline .idea/inspectionProfiles/Project_Default.xml:1]: It looks like your editor added these files.  You may want to add them to the `.gitignore` file.

IntelliJ has documentation on which of its project settings files can be checked into source control (so a team can share the same settings) versus which files should be ignored (for settings that are unique to a developer's workstation configuration. e.g. absolute filesystem paths to project files, etc.)

* https://intellij-support.jetbrains.com/hc/en-us/articles/206544839-How-to-manage-projects-under-Version-Control-Systems
* https://www.jetbrains.com/help/idea/configure-project-settings.html

File: app/src/main/java/com/ikarabulut/App.java [hunk 2]
```diff
@@ -3,12 +3,23 @@
...
-    public String getGreeting() {
-        return "Hello World!";
-    }
+import java.io.IOException;
+import java.net.ServerSocket;
 
+public class App {
     public static void main(String[] args) {
-        System.out.println(new App().getGreeting());
+        int port = 5000;
```
Comment (riccjohn) [inline app/src/main/java/com/ikarabulut/App.java:11]: If possible in Java, this is the sort of variable I'd define outside the class. For now, you can try putting it above the class definition as `int PORT = 5000`. Usually when you see an all-caps var, it means it's something that won't change and is more configuration.

Maybe later on, we can try getting something like that moved to a .env file

File: app/src/test/java/com/ikarabulut/ServerIOTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        PrintWriter printWriter = new PrintWriter(out, true);
+
+        io.printOutput(printWriter, inputStream);
+
+        assertEquals("1\n", out.toString());
+    }
+
+    @Test
+    @DisplayName("When a socket writer gets created, then null should not be returned")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/ServerIOTest.java:62]: What _should_ get returned?

File: app/src/test/java/com/ikarabulut/ServerIOTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        ServerIO io = new ServerIO();
+        Socket mockSocket = mock(Socket.class);
+        OutputStream mockOutputStream = mock(OutputStream.class);
+        when(mockSocket.getOutputStream()).thenReturn(mockOutputStream);
+
+        assertNotNull(io.generateClientSocketWriter(mockSocket));
+    }
+
+    @Test
+    @DisplayName("When a socket reader gets created, then null should not be returned")
```
Comment (riccjohn) [inline app/src/test/java/com/ikarabulut/ServerIOTest.java:73]: Same as above comment

**Review by riccjohn — CHANGES_REQUESTED** [review]

File: app/src/test/java/com/ikarabulut/ServerIOTest.java (hunk 3 above)
Comment (jramnani) [reply app/src/test/java/com/ikarabulut/ServerIOTest.java:62]: Right. Describing what _shouldn't_ be returned is helpful when I'm in a tight TDD loop ("Does not return null means I'm creating the object, now on to adding the next behavior."), but is not useful to the future teammate or other code reader once I'm outside that workflow. After all, the things this function _shouldn't_ return is infinite. It should not return null, but it should also not return the number 123, or the String, "hello". 

When I'm on a client project, there is an editing step where I review the code before submitting it to my teammates, where I try to put myself into their shoes and think about how to describe what I've done to them.  I know I've failed that step when I say something like "it just does the thing, you know".

What should be returned? Sounds like a PrintStream? Or, rather than that concrete class name, what behavior does it provide to other objects in the program?

Comment (jramnani) [reply app/src/test/java/com/ikarabulut/ServerIOTest.java:62]: Put another way: when naming or describing tests, prefer a positive statement rather than a negative one.

Rather than "does not return null", or "does now allow invalid data", which don't tell the reader anything useful, prefer positive statements like these examples: "input must not be below zero", "input must not be above 100", or "input must be between 0 and 100", or "returns an object that can be written to".

File: app/src/main/java/com/ikarabulut/App.java [hunk 5]
```diff
@@ -3,12 +3,23 @@
...
+        ServerIO io = new ServerIO();
+
+        try {
+            ServerSocket serverSocket = new ServerSocket(port);
+            System.out.println("You are connected to port: " + port);
+            EchoServer echoServer = new EchoServer(serverSocket);
+            echoServer.createServerEndPoint(io);
```
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/App.java:18]: Passing it through the constructor is a better route to go. It helps with threading as well. In the multithreaded branch I am messing around with, I ended up putting it in the constructor to get it working. So ill update it here

File: app/src/test/java/com/ikarabulut/ServerIOTest.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    void readInputStream_WithHello() {
+        ServerIO io = new ServerIO();
+        String inputStream = "Hello";
+        BufferedReader input = new BufferedReader(new InputStreamReader(new ByteArrayInputStream(inputStream.getBytes())));
+
+        assertEquals("Hello", io.readInput(input));
+    }
+
+    @Test
+    @DisplayName("When an input of 1 is passed, then readInputStream should return 1")
```
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/ServerIOTest.java:26]: My initial thought was to pass an integer but then realized that a string is always passed anyway so it isn't a necessary test but it was already typed, so i figured i'd just leave it

File: app/src/test/java/com/ikarabulut/EchoServerTest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+    @Test
+    @DisplayName("When a server socket accepts requests, a client socket will be returned")
+    void createClientSocket() throws IOException {
+        ServerSocket mockServerSocket = mock(ServerSocket.class);
+        EchoServer echoServer = new EchoServer(mockServerSocket);
+
+        when(mockServerSocket.accept()).thenReturn(new Socket());
+
+        assertNotNull(echoServer.createClientSocket(mockServerSocket));
```
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/test/java/com/ikarabulut/EchoServerTest.java:33]: Yes, I updated this assertion to first create a local variable `newClientSocket` which is set to a new Client object. When the mock runs `.accept` it will return the `newClientSocket` and I am asserting that the returned value is the same as `newClientSocket`

File: app/src/main/java/com/ikarabulut/ServerIO.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import java.net.Socket;
+
+public class ServerIO {
+    public String readInput(BufferedReader input) {
+        try {
+            return input.readLine();
+        } catch (IOException ex) {
+            System.err.print(ex.getMessage());
+        }
+        return null;
```
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/ServerIO.java:16]: That second comment makes sense. I haven't thought that deep into exceptions and now they make a bit more sense to me. 

To answer your question, Based on this StackOverflow it looks like an IO Error is very unlikely to occur on `readLine()` https://stackoverflow.com/questions/29202584/when-might-an-ioerror-be-thrown 
but if it were to occur then it most likely could not recover from that error scenario.

I'll update this to just throw the exception and as the caller to handle it.

File: app/src/main/java/com/ikarabulut/EchoServer.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    private ServerSocket serverSocket;
+    private PrintWriter echoWriter;
+    private BufferedReader echoReader;
+
+    public EchoServer(ServerSocket connectedServerSocket) {
+        serverSocket = connectedServerSocket;
+    }
+
+    public void createServerEndPoint(ServerIO io) throws IOException {
+        clientSocket = createClientSocket(serverSocket);
```
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/EchoServer.java:20]: So I ran into this issue when I was trying multithreading. The terminal opens up an input stream and I can pass something but then it just hangs. So the program did not crash but it also didn't create the EndPoint connection but did not let me know of that.

I updated `createClientSocket` to not return `null` and reran the server and it still does the same when I try to open two terminal windows running localhost 5000. 

I guess I can find a way to let the user know that only one connection is required. But with multithreading this wont be an issue. It would be cool to show the client how many connections the server has and which number they are

**Review by riccjohn — COMMENTED** [review]
Should the .idea/ dir be gitignore'd now?

File: app/src/test/java/com/ikarabulut/ServerIOTest.java (hunk 3 above)
Comment (riccjohn) [reply app/src/test/java/com/ikarabulut/ServerIOTest.java]: Re: your comment in Slack -> https://8thlight.slack.com/archives/C033HEKJ2LC/p1654011654770819

You're right that the types are doing making some of the assurances here, but I think you focus is on what _thing_ (class) gets returned, rather than _how_ that class instance gets created. For example, rather than focusing on if `null` does not get returned or if a `BufferedReader` does get returned, why don't we test that when `getClientSocketWriter` gets called with a `Socket`, we expect a new `PrintWriter` to get created and we expect it to receive the correct data.

You're doing in one line what I might do in two, and that might be hiding some of the complexity that I think you should test.

```
  public PrintWriter generateClientSocketWriter(Socket clientSocket) throws IOException {
      return new PrintWriter(clientSocket.getOutputStream(), true);
  }
```

```
  public PrintWriter generateClientSocketWriter(Socket clientSocket) throws IOException {
      outputStream = clientSocket.getOutputStream();
      return new PrintWriter(outputStream, true);
  }
  ```
  
  Now, I think it is a bit clearer what we want to test. `generateClientSocketWriter` will receive a `Socket` and use that (and another value) to create a new `PrintWriter`. Tests can be:
  
  - expect a new PrintWriter to be created
  - expect a new PrintWriter to be crated with the outputStream of the givenSocket, and the value `true`
  
 You may also want to write a test that says something about the error that's thrown. Something like "when X happens, it will throw an error of type Y"

**Comment by ikarabulut** [comment]
So I think that was a mistake I did. I followed these steps https://www.atlassian.com/git/tutorials/saving-changes/gitignore#ignoring-a-previously-committed

And did remove them from the git cache but then ran rm .idea which deleted the directory locally, and then committed the changes so it looks like it pulled in the deletion after the cache was cleared.

But moving forward there will be no .idea files in my repo. The cache for them is currently empty, I just checked

File: .idea/inspectionProfiles/Project_Default.xml (hunk 1 above)
Comment (jramnani) [reply .idea/inspectionProfiles/Project_Default.xml]: It's probably best to ignore the `.idea` directory for now.

File: app/src/main/java/com/ikarabulut/EchoServer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jramnani) [reply app/src/main/java/com/ikarabulut/EchoServer.java]: > reran the server and it still does the same when I try to open two terminal windows running localhost 5000.

That's because it's not up to your program to proceed. The operating system will not allow multiple programs to bind to the same port on the same machine. This error is not recoverable by a user program.  Other than to have two servers listening on different ports.

File: .idea/inspectionProfiles/Project_Default.xml (hunk 1 above)
Comment (ikarabulut) [reply .idea/inspectionProfiles/Project_Default.xml]: Directory has been ignored

File: app/src/main/java/com/ikarabulut/App.java (hunk 2 above)
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/App.java]: done

File: app/src/main/java/com/ikarabulut/EchoServer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (ikarabulut) [reply app/src/main/java/com/ikarabulut/EchoServer.java]: Oh ok, I understand this now. Only one server will be listening or accepting requests from that port. 

So in this iteration, `createServerEndPoint` should take in the `clientSocket` as a parameter and then use that to generate the reader and writer. Since the running server will only run `.accept` once and then just take in requests.

My new iteration of the server took a different approach to getting this done so the method is outdated but I understand what was happening here.
